classes/ValueFunction.py

```diff
@@ -0,0 +1,146 @@
+import classes
+import itertools
+from globals import *
+import numpy as np
+
+
+class ValueFunction:
+    def __init__(
+        self,
+        number_of_clusters: int,
+        number_of_depots=3,
+        weight_update_step_size=0.1,
+        discount_factor=DISCOUNT_RATE,
+        vehicle_inventory_step_size=0.25,
+        number_of_small_depots=2,
+        number_of_features_per_cluster=2,
+    ):
+        # for every location - 3 bit for location
+        # for every cluster 1 float for deviation, 1 float for battery deficient
+        # for vehicle - n bits for scooter inventory in percentage ranges (e.g 0-10%, 10%-20%, etc..)
+        # + n bits for battery inventory in percentage ranges (e.g 0-10%, 10%-20%, etc..)
+        # for every small depot - 1 float for degree of filling
+        number_of_locations_features = (number_of_clusters + number_of_depots) * 3
+        number_of_state_features = (
+            (number_of_features_per_cluster * number_of_clusters)
+            + (2 * int(1 / vehicle_inventory_step_size))
+            + number_of_small_depots
+        )
+
+        self.weights = [0] * (
+            number_of_locations_features
+            + number_of_state_features
+            + (number_of_locations_features * number_of_state_features)
+        )
+        self.location_indicator = [0] * (3 * (number_of_clusters + number_of_depots))
+        self.vehicle_inventory_step_size = vehicle_inventory_step_size
+        self.step_size = weight_update_step_size
+        self.discount_factor = discount_factor
+
+    def estimate_value(
+        self, state: classes.State, vehicle: classes.Vehicle, time: int,
+    ):
+
+        current_linear_model_features = ValueFunction.create_location_features_combination(
+            self.convert_state_to_features(state, vehicle, time)
+        )
+
+        current_state_value = float(np.dot(current_linear_model_features, self.weights))
+
+        return current_state_value
+
+    def update_weights(
+        self,
+        current_state_features: [float],
+        current_state_value: float,
+        next_state_value: float,
+        reward: float,
+    ):
+        self.weights += (
+            self.step_size
+            * (reward - (self.discount_factor * next_state_value) - current_state_value)
+            * current_state_features
+        )
+
+    def convert_state_to_features(
+        self, state: classes.State, vehicle: classes.Vehicle, time: int
+    ):
+        location_indicator = (
+            self.location_indicator[: vehicle.current_location.id * 3]
+            + [1, 1, 1]
+            + self.location_indicator[(vehicle.current_location.id + 1) * 3 :]
+        )
```

Comment:
Do you need self.location_indicator?
```suggestion
        location_indicator = [0]*3*(vehicle.current_location.id-1)+[1]*3+[0]*3*(number_of_clusters + number_of_depots-pos)
```

Reply:
Copy